File: medeia-validator-java-examples/src/main/java/com/worldturner/medeia/examples/java/gson/customformats/CustomFormatExample.java
package com.worldturner.medeia.examples.java.gson.customformats;

import com.google.gson.stream.JsonReader;
import com.worldturner.medeia.api.*;
import com.worldturner.medeia.api.gson.MedeiaGsonApi;
import com.worldturner.medeia.schema.validation.SchemaValidator;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.IOException;
import java.io.InputStreamReader;
import java.io.Reader;
import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Locale;
import java.util.Map;

public class CustomFormatExample {
    static class PalindromeValidator implements FormatValidation {
        @Nullable
        @Override
        public String validate(@Nullable Object value, @NotNull String format) {
            String text = String.valueOf(value);
            String nospaces = text.replaceAll("\\s+", "");
            String lowercase = nospaces.toLowerCase(Locale.US);
            StringBuffer reversed = new StringBuffer(lowercase);
            reversed.reverse();
            if (!lowercase.equals(reversed.toString())) {
                return "not a palindrome";
            }
            return null;
        }
    }

    private static MedeiaGsonApi api = new MedeiaGsonApi();

    private void parseValidExample() throws IOException {
        SchemaValidator validator = loadSchema();
        Reader dataReader =
                new InputStreamReader(
                        getClass().getResourceAsStream("/customformats/valid-data.json"),
                        StandardCharsets.UTF_8);
        JsonReader validatedReader = api.createJsonReader(validator, dataReader);
        api.parseAll(validatedReader);
    }

    private void parseInvalidExample() throws IOException {
        SchemaValidator validator = loadSchema();
        Reader dataReader =
                new InputStreamReader(
                        getClass().getResourceAsStream("/customformats/invalid-data.json"),
                        StandardCharsets.UTF_8);
        JsonReader validatedReader = api.createJsonReader(validator, dataReader);
        try {
            api.parseAll(validatedReader);
            throw new IllegalStateException("Invalid json data passed validation");
        } catch (ValidationFailedException e) {
            // Expected
            System.out.println("Validation failed as expected: " + e);
        }
    }

    @NotNull
    private SchemaValidator loadSchema() {
        SchemaSource source = new UrlSchemaSource(
                getClass().getResource("/customformats/customformats-schema.json"));
        Map<String, FormatValidation> customFormats = new HashMap<>();
        customFormats.put("palindrome", new PalindromeValidator());
        return api.loadSchemas(
                Arrays.asList(source),
                new ValidationOptions().withCustomFormats(customFormats));
    }

    public static void main(String[] args) throws IOException {
        CustomFormatExample example = new CustomFormatExample();
        example.parseValidExample();
        example.parseInvalidExample();
    }
}

File: medeia-validator-kotlin-examples/src/main/kotlin/com/worldturner/medeia/examples/kotlin/gson/customformats/CustomFormatExample.kt
package com.worldturner.medeia.examples.kotlin.gson.customformats

import com.worldturner.medeia.api.FormatValidation
import com.worldturner.medeia.api.ValidationOptions
import com.worldturner.medeia.api.UrlSchemaSource
import com.worldturner.medeia.api.ValidationFailedException
import com.worldturner.medeia.api.gson.MedeiaGsonApi
import com.worldturner.medeia.schema.validation.SchemaValidator
import java.io.InputStreamReader
import java.util.Locale

class CustomFormatExample {
    private val api = MedeiaGsonApi()

    fun parseValidExample() {
        val validator = loadSchema()
        val dataReader = InputStreamReader(
            javaClass.getResourceAsStream("/customformats/valid-data.json"),
            Charsets.UTF_8
        )
        val validatedReader = api.createJsonReader(validator, dataReader)
        api.parseAll(validatedReader)
    }

    fun parseInvalidExample() {
        val validator = loadSchema()
        val dataReader = InputStreamReader(
            javaClass.getResourceAsStream("/customformats/invalid-data.json"),
            Charsets.UTF_8
        )
        val validatedReader = api.createJsonReader(validator, dataReader)
        try {
            api.parseAll(validatedReader)
            throw IllegalStateException("Invalid json data passed validation")
        } catch (e: ValidationFailedException) {
            // Expected
            println("Validation failed as expected: $e")
        }
    }

    private fun loadSchema(): SchemaValidator {
        val source = UrlSchemaSource(
            javaClass.getResource("/customformats/customformats-schema.json")
        )
        val customFormats = mapOf("palindrome" to PalindromeValidator())
        return api.loadSchemas(listOf(source), ValidationOptions(customFormats = customFormats))
    }
}

class PalindromeValidator : FormatValidation {
    override fun validate(value: Any?, format: String): String? {
        val text = value.toString()
        val nospaces = text.replace("\\s+".toRegex(), "")
        val lowercase = nospaces.toLowerCase(Locale.US)
        val reversed = StringBuffer(lowercase).also { it.reverse() }.toString()
        return if (lowercase != reversed) "not a palindrome" else null
    }
}

fun main(args: Array<String>) {
    val example = CustomFormatExample()
    example.parseValidExample()
    example.parseInvalidExample()
}